diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0446f24..2cb88ab 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -34,7 +34,7 @@ pkg_check_modules(ZSTD REQUIRED IMPORTED_TARGET "libzstd")
 find_package(Fcitx5Utils REQUIRED)
 include("${FCITX_INSTALL_CMAKECONFIG_DIR}/Fcitx5Utils/Fcitx5CompilerSettings.cmake")
 
-find_package(Boost 1.61 REQUIRED COMPONENTS iostreams filesystem)
+find_package(Boost 1.61 REQUIRED COMPONENTS iostreams)
 set(LIBIME_INSTALL_PKGDATADIR "${CMAKE_INSTALL_FULL_DATADIR}/libime")
 set(LIBIME_INSTALL_LIBDATADIR "${CMAKE_INSTALL_FULL_LIBDIR}/libime")
 
diff --git a/src/libime/core/CMakeLists.txt b/src/libime/core/CMakeLists.txt
index 89dd191..ed3f1b5 100644
--- a/src/libime/core/CMakeLists.txt
+++ b/src/libime/core/CMakeLists.txt
@@ -68,7 +68,7 @@ ecm_setup_version(PROJECT
 # the cmake if will
 set(IMECore_SOVERSION 0)
 
-add_library(IMECore SHARED ${LIBIME_SRCS})
+add_library(IMECore STATIC ${LIBIME_SRCS})
 set_target_properties(IMECore
                       PROPERTIES VERSION ${IMECore_VERSION}
                       SOVERSION ${IMECore_SOVERSION}
@@ -81,7 +81,7 @@ target_include_directories(IMECore PUBLIC
 
 target_link_libraries(IMECore PUBLIC Fcitx5::Utils Boost::boost PRIVATE kenlm Boost::iostreams PkgConfig::ZSTD)
 
-install(TARGETS IMECore EXPORT LibIMECoreTargets LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}" COMPONENT lib)
+install(TARGETS IMECore kenlm EXPORT LibIMECoreTargets LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}" COMPONENT lib)
 install(FILES ${LIBIME_HDRS} DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/LibIME/libime/core" COMPONENT header)
 
 add_library(LibIME::Core ALIAS IMECore)
diff --git a/src/libime/pinyin/CMakeLists.txt b/src/libime/pinyin/CMakeLists.txt
index 4da3068..d951860 100644
--- a/src/libime/pinyin/CMakeLists.txt
+++ b/src/libime/pinyin/CMakeLists.txt
@@ -30,7 +30,7 @@ ecm_setup_version(PROJECT
                   PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/LibIMEPinyinConfigVersion.cmake")
 set(IMEPinyin_SOVERSION 0)
 
-add_library(IMEPinyin SHARED ${LIBIME_PINYIN_SRCS})
+add_library(IMEPinyin STATIC ${LIBIME_PINYIN_SRCS})
 set_target_properties(IMEPinyin PROPERTIES
                       VERSION ${IMEPinyin_VERSION}
                       SOVERSION ${IMEPinyin_SOVERSION}
diff --git a/src/libime/table/CMakeLists.txt b/src/libime/table/CMakeLists.txt
index 62317bd..dd47d0b 100644
--- a/src/libime/table/CMakeLists.txt
+++ b/src/libime/table/CMakeLists.txt
@@ -23,7 +23,7 @@ ecm_setup_version(PROJECT
                   PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/LibIMETableConfigVersion.cmake")
 set(IMETable_SOVERSION 0)
 
-add_library(IMETable SHARED ${LIBIME_TABLE_SRCS})
+add_library(IMETable STATIC ${LIBIME_TABLE_SRCS})
 set_target_properties(IMETable
                       PROPERTIES VERSION ${IMETable_VERSION}
                       SOVERSION ${IMETable_SOVERSION}
diff --git a/tools/CMakeLists.txt b/tools/CMakeLists.txt
index ca1e6ac..a214977 100644
--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -23,13 +23,3 @@ add_executable(libime_tabledict libime_tabledict.cpp)
 target_link_libraries(libime_tabledict LibIME::Table)
 install(TARGETS libime_tabledict DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT tools)
 add_executable(LibIME::tabledict ALIAS libime_tabledict)
-
-add_executable(libime_migrate_fcitx4_table libime_migrate_fcitx4_table.cpp)
-target_link_libraries(libime_migrate_fcitx4_table LibIME::Table Boost::iostreams Boost::filesystem)
-install(TARGETS libime_migrate_fcitx4_table DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT tools)
-add_executable(LibIME::migrate_fcitx4_table ALIAS libime_migrate_fcitx4_table)
-
-add_executable(libime_migrate_fcitx4_pinyin libime_migrate_fcitx4_pinyin.cpp)
-target_link_libraries(libime_migrate_fcitx4_pinyin LibIME::Pinyin Boost::iostreams Boost::filesystem)
-install(TARGETS libime_migrate_fcitx4_pinyin DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT tools)
-add_executable(LibIME::migrate_fcitx4_pinyin ALIAS libime_migrate_fcitx4_pinyin)
